modules/os-sso73/added/standalone-openshift.xml

```diff
@@ -0,0 +1,661 @@
+            <theme>
+                <staticMaxAge>2592000</staticMaxAge>
+                <cacheThemes>true</cacheThemes>
+                <cacheTemplates>true</cacheTemplates>
+                <dir>${jboss.home.dir}/themes</dir>
+            </theme>
+            <spi name="eventsStore">
+                <provider name="jpa" enabled="true">
+                    <properties>
+                        <property name="exclude-events" value="[&quot;REFRESH_TOKEN&quot;]"/>
+                    </properties>
+                </provider>
+            </spi>
+            <spi name="userCache">
+                <provider name="default" enabled="true"/>
+            </spi>
+            <spi name="userSessionPersister">
+                <default-provider>jpa</default-provider>
+            </spi>
+            <spi name="timer">
+                <default-provider>basic</default-provider>
+            </spi>
+            <spi name="connectionsHttpClient">
+                <provider name="default" enabled="true"/>
+            </spi>
+            <spi name="connectionsJpa">
+                <provider name="default" enabled="true">
+                    <properties>
+                        <property name="dataSource" value="java:jboss/datasources/KeycloakDS"/>
+                        <property name="initializeEmpty" value="true"/>
+                        <property name="migrationStrategy" value="update"/>
+                        <property name="migrationExport" value="${jboss.home.dir}/keycloak-database-update.sql"/>
+                    </properties>
+                </provider>
+            </spi>
+            <spi name="realmCache">
+                <provider name="default" enabled="true"/>
+            </spi>
+            <spi name="connectionsInfinispan">
+                <default-provider>default</default-provider>
+                <provider name="default" enabled="true">
+                    <properties>
+                        <property name="cacheContainer" value="java:comp/env/infinispan/Keycloak"/>
+                    </properties>
+                </provider>
+            </spi>
+            <spi name="jta-lookup">
+                <default-provider>${keycloak.jta.lookup.provider:jboss}</default-provider>
+                <provider name="jboss" enabled="true"/>
+            </spi>
+            <spi name="publicKeyStorage">
+                <provider name="infinispan" enabled="true">
+                    <properties>
+                        <property name="minTimeBetweenRequests" value="10"/>
+                    </properties>
+                </provider>
+            </spi>
+            <spi name="x509cert-lookup">
+                <default-provider>${keycloak.x509cert.lookup.provider:default}</default-provider>
+                <provider name="default" enabled="true"/>
+            </spi>
+            <spi name="hostname">
+                <default-provider>fixed</default-provider>
```

Comment:
This is not going to work. It will require setting the HOSTNAME variable always, which is not what we want.

The behaviour we want is:

* By default its not needed to specify hostname. The config required here is <default-provider>request</default-provider>
* Have a single env variable to set HOSTNAME. This is highly recommended in production, but not needed in dev/testing usage unless you're doing backchannel requests with a different URL. Config required here is: <default-provider>fixed</default-provider> and <property name="hostname" value="HOSTNAME"/>

The way this is implemented in community is:

https://github.com/jboss-dockerfiles/keycloak/blob/master/server/tools/cli/hostname.cli
https://github.com/jboss-dockerfiles/keycloak/blob/master/server/tools/docker-entrypoint.sh#L15

The config in standalone.xml will look like:

<default-provider>${keycloak.hostname.provider:request}</default-provider>
...
<property name="hostname" value="${keycloak.hostname.fixed.hostname:localhost}"/>

The startup script takes care of setting the required system variables based if $KEYCLOAK_HOSTNAME is set, like this:

if [ "$KEYCLOAK_HOSTNAME" != "" ]; then
    SYS_PROPS="-Dkeycloak.hostname.provider=fixed -Dkeycloak.hostname.fixed.hostname=$KEYCLOAK_HOSTNAME"
fi

Reply:
@stianst 
Thanks, Stian! Right, realized this while testing the current image yesterday evening. Seems the Keycloak server hostname needs to match `host:` field in the route definition. If not (like in this case), clicking the route URL is redirected to the Keycloak server hostname, which leads to `application is not available or is just starting message`.

Simultaneously, I don't want to expose the new `SSO_HOSTNAME` variable as a new parameter (possibly containing arbitrary host name) in the templates. While this would work for local OCP installs, it wouldn't for Online environment [CLOUD-2216](https://issues.jboss.org/browse/CLOUD-2216) / [KEYCLOAK-7936](https://issues.jboss.org/browse/KEYCLOAK-7396), since it's not possible to set `host:` field of route on Online to arbitrary URL. If the user set arbitrary `SSO_HOSTNAME` URL for Online, they would hit the redirect issue again.

But what could work (needs testing yet though), is to pass new `SSO_HOSTNAME` variable to RH-SSO pod (defaulting to same `host:` setting, like route has), and in this case the redirection would work, and still the user wouldn't be able to specify arbitrary URL, leading to setup not working on Online.

And yet, if the image was launched directly (without the template), and `SSO_HOSTNAME` wouldn't be set, we would issue a warning `Please specify SSO_HOSTNAME variable` and exit.

Will know more, once modified `standalone-openshift.xml` per your suggestion above, and got chance to test the modified templates.